src/core/lib/event_engine/uv/libuv_event_engine.cc

```diff
@@ -0,0 +1,427 @@
+          gpr_log(GPR_DEBUG, "LibuvTask@%p, triggered: key = %" PRIiPTR, task,
+                  task->Key());
+        }
+        task->ran_ = true;
+        // TODO(hork): Timer callbacks will be delayed by one iteration of the
+        // uv_loop to avoid race conditions around EventEngine destruction.
+        // Before the timer callback has run, the uv state for that timer is
+        // destroyed. This is delay is not ideal, we should find a way to avoid
+        // it.
+        uv_close(reinterpret_cast<uv_handle_t*>(timer),
+                 &LibuvTask::RunAndErase);
+      },
+      timeout, 0);
+}
+
+void LibuvTask::Cancel(Promise<bool>& will_be_cancelled) {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+    gpr_log(GPR_DEBUG, "LibuvTask@%p, cancelled: key = %" PRIiPTR, this, key_);
+  }
+  if (uv_is_closing(reinterpret_cast<uv_handle_t*>(&timer_)) != 0) {
+    // TODO(hork): check if this can be called on uv shutdown instead
+#ifndef NDEBUG
+    GPR_ASSERT(GPR_LIKELY(ran_));
+#endif
+    will_be_cancelled.Set(false);
+    return;
+  }
+  will_be_cancelled.Set(true);
+  uv_timer_stop(&timer_);
+  uv_close(reinterpret_cast<uv_handle_t*>(&timer_), &LibuvTask::Erase);
+}
+
+void LibuvTask::Erase(uv_handle_t* handle) {
+  uv_timer_t* timer = reinterpret_cast<uv_timer_t*>(handle);
+  LibuvTask* task = reinterpret_cast<LibuvTask*>(timer->data);
+  LibuvEventEngine* engine =
+      reinterpret_cast<LibuvEventEngine*>(timer->loop->data);
+  engine->EraseTask(task->key_);
+}
+
+void LibuvTask::RunAndErase(uv_handle_t* handle) {
+  uv_timer_t* timer = reinterpret_cast<uv_timer_t*>(handle);
+  LibuvTask* task = reinterpret_cast<LibuvTask*>(timer->data);
+  std::function<void()> fn = std::move(task->fn_);
+  LibuvEventEngine* engine =
+      reinterpret_cast<LibuvEventEngine*>(timer->loop->data);
+  engine->EraseTask(task->key_);
+  fn();
+}
+
+LibuvEventEngine::LibuvEventEngine() {
+  bool success = false;
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+    gpr_log(GPR_DEBUG, "LibuvEventEngine:%p created", this);
+  }
+  thread_ = grpc_core::Thread(
+      "uv loop",
+      [](void* arg) {
+        LibuvEventEngine* engine = reinterpret_cast<LibuvEventEngine*>(arg);
+        engine->RunThread();
+      },
+      this, &success);
+  thread_.Start();
+  GPR_ASSERT(GPR_LIKELY(success));
+  // This promise will be set to true once the thread has fully started and is
+  // operational, so let's wait on it.
+  success = ready_.Get();
+  GPR_ASSERT(GPR_LIKELY(success));
+}
+
+LibuvEventEngine::~LibuvEventEngine() {
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+    gpr_log(GPR_DEBUG, "LibuvEventEngine@%p::~LibuvEventEngine", this);
+  }
+  RunInLibuvThread([](LibuvEventEngine* engine) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_tcp_trace)) {
+      gpr_log(GPR_DEBUG,
+              "LibuvEventEngine@%p shutting down, unreferencing Kicker now",
+              engine);
+    }
+    GPR_ASSERT(engine->uv_shutdown_can_proceed_.Get());
```

Comment:
This promise seems now to be limited to that destructor. It can be moved away from the class definition itself and into the stack of the destructor instead. You'd need to capture it, of course.

Reply:
Good catch. Moved and captured.